### PR TITLE
feat(firezone-tunnel): sort resources alphabetically

### LINKED ISSUE #3217: Sort resources by name alphabetically before calling `onUpdateResources`
**Describe the bug**
Resources are in random order. If you close the client and restart it, they shuffle again.

**To Reproduce**
Open the client and sign in, close it, open it again.

**Expected behavior**
I expect them to be ordered alphabetically, and maybe grouped by site. I'm not sure what the other clients do.

**Screenshots / Logs**
image

image

**Platform (please complete the following information):**
 - Firezone Version 04aeee2e7bf84
 - OS: Windows 11

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by ReactorScram**
Closes #3217. I just now noticed that one was assigned to me

image

The sorting is naive, just sorts the UTF-8 encoded bytes, so lowercase resources come after all uppercase resources, and it's probably very wrong for anything outside Latin-1 and English locale. If the names are identical, resource ID tie-breaks.

### CODE REVIEW

**Review by jamilbk — COMMENTED** [review]
This should suffice in the short-term until we decide to bubble up site names from the portal to group resources by.

Will defer to @thomaseizinger or @conectado for the Rust review

**Comment by jamilbk** [comment]
Actually, now that I think about it, a better key UX-wise to sort these Resources by would be `created_at ASC` from the portal. That way we won't be bumping around the order of Resources since I'm sure some users will be clicking those a lot

But I think this will require that field to be rendered from the portal? And appropriate type added to connlib. Can save for later.

**Comment by AndrewDryga** [comment]
@jamilbk Bubbling is implemented in #3466 but UX needs to be thoughtful around multi-site resources (how we should group them), I'd suggest something like this with resources ordered alphabetically within groups:
```
- Multi-site resource
- Multi-site resource
--- Site A ---
- Site A resource
- Site A resource
--- Site B ---
- Site B resource
```

File: rust/connlib/tunnel/src/client.rs [hunk 1]
```diff
@@ -83,12 +83,22 @@ where
...
+                .insert(resource_description.id(), resource_description);
+            role_state
+                .resource_ids
+                .values()
+                .cloned()
+                .collect::<Vec<_>>()
+        };
+        // Unstable sort is slightly faster, and we use the ID as a tie-break
+        // if the display name is identical for some reason, so it should still be stable overall
+        resource_descriptions.sort_unstable_by(|a, b| (a.name(), a.id()).cmp(&(b.name(), b.id())));
```
Comment (thomaseizinger) [inline rust/connlib/tunnel/src/client.rs]: Can we extract this into a function and add 1 or 2 unit tests?

**Review by thomaseizinger — APPROVED** [review]

**Comment by thomaseizinger** [comment]
> Actually, now that I think about it, a better key UX-wise to sort these Resources by would be `created_at ASC` from the portal. That way we won't be bumping around the order of Resources since I'm sure some users will be clicking those a lot
> 
> But I think this will require that field to be rendered from the portal? And appropriate type added to connlib. Can save for later.

I think this can be even more confusing. What if an admin needs to re-create a resource?

I think alphabetically is a good default for connlib. Later we can have clients offer an override where a user can sort them in the GUI which will then take precedence.

**Comment by jamilbk** [comment]
> What if an admin needs to re-create a resource?

Yeah, this is why I'm still an advocate for #2504. If an admin is recreating existing Resources when the real-world asset it references hasn't materially changed I'd call that a usability issue. Today we just have a `name` field, but eventually they'll have to duplicate more things like traffic filters, description field, etc.

Infosec users viewing audit reports (#949) will need to correlate access logs across multiple "duplicated" (to them) Resources. Exporting to their SIEM will generate many resource IDs per real-world asset. It'll be harder to write SIEM queries for. It'd be easier if we kept one UUID per real-world asset as much as possible in order to answer questions like "how many times did this user access the engineering GitLab?" even if it happens to change IPs or DNS names over its lifetime.

This works for now, and when we update clients to show the Sites grouping we can debate further then. This will probably be a heated discussion so we can use our new standup timing ;-). Hopefully we'll have some user feedback to seed the discussion.

**Pull request merged**